cpp/src/parquet/util/memory.cc

```diff
@@ -270,7 +270,7 @@ const uint8_t* BufferedInputStream::Peek(int64_t num_to_peek, int64_t* num_bytes
     DCHECK(buffer_size_ >= *num_bytes);
   }
   // Read more data when buffer has insufficient left or when resized
-  if (*num_bytes > (buffer_size_ - buffer_offset_)) {
+  if (*num_bytes >= (buffer_size_ - buffer_offset_)) {
```

Comment:
I don't think that's the problem. You only need to read more data if you want strictly more than what's available in the buffer (so `>`).

There are other problems though:
- there is a confusion with `buffer_size_`, which claims to track the number of bytes available, but we change it in `Resize` above without reading more bytes; really, the code should distinguish between buffer size and capacity
- `Peek()` ignores any pending data in the buffer if `num_to_peek` is larger than the buffer size